### ISSUE TITLE
Will you upload your training code?

### ISSUE THREAD

**Issue by SKBL5694** [mamonraab/fire-detection-in-video-with-pytroch#5]
Thanks for you article and detailed explanation, it's very helpful. It's sorry that I am a beginner in pytorch, I can't make it train easily. I found that the issue #3 talked about some code which maybe useful for me, but it seems still have some problems to solve yet. So can you upload your train code if it's convenient or I just use the code in #3 ?
Thanks again for your great work and video explanation.
regards

**Comment by mamonraab** [comment]
it is easy task to be done but since you asked for that and others also asking for training code ,
give me few days to find free time and will upload easy to fellow training code

**Comment by SKBL5694** [comment]
Sorry for my poor coding skills, thanks again and looking forward to your code.

**Comment by jwhmicrsoft** [comment]
Hi, I got ur messge. Thanks to the @mamonraab 's code I complete my job. I will show you my code later.

**Comment by jwhmicrsoft** [comment]
btw, I found that cv2.resize is much better than the skimage's resize which makes the training much much slower. I highly recommend replace it by cv2.resize.
here is my code：
### Loading the video data
```
class VideoDataset(Dataset):

    def __init__(self, datas, timestep=28, rgb=3, h=120, w=180):
        self.datalocation = datas
        self.timesep, self.rgb, self.h, self.w = timestep, rgb, h, w

    def __len__(self):
        return len(self.datalocation)

    def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()
        video = capture(self.datalocation.iloc[idx, 0], self.timesep, self.rgb, self.h, self.w)
        # sample = {'video': torch.from_numpy(video), 'label': self.datalocation.iloc[idx, 1]}

        return torch.from_numpy(video), self.datalocation.iloc[idx, 1]
```
### Capture and resize the frame of the video
```

def capture(filename, timesep, rgb, h, w):
    tmp = []
    frames = np.zeros((timesep, rgb, h, w), dtype=np.float32)
    i = 0
    vc = cv2.VideoCapture(filename)
    frame = np.zeros((1, h, w), dtype=np.float32)
    frame_all = vc.get(cv2.CAP_PROP_FRAME_COUNT) 
    # frames = np.zeros((frame_all, rgb, h, w), dtype=np.float32) 
    if vc.isOpened():
        rval, frame = vc.read()

    frm = cv2.resize(frame, (w, h))
    frm = np.expand_dims(frm, axis=0)
    frm = np.moveaxis(frm, -1, 1)

    if (np.max(frm) > 1):
        frm = frm / 255.0
    frames[i][:] = frm
    i += 1

    while i < timesep and i < frame_all:
        tmp[:] = frm[:]
        rval, frame = vc.read()
        frm = cv2.resize(frame, (w, h))
        frm = np.expand_dims(frm, axis=0)
        if np.max(frm) > 1:
            frm = frm / 255.0
        frm = np.moveaxis(frm, -1, 1)
        frames[i - 1][:] = frm  # - tmp
        i += 1
    return frames

```

**Comment by jwhmicrsoft** [comment]
Btw, I wrote a csv maker and a video slicer, hope you will find it helpful.

### Video Slicer
```

import cv2
import os

vid_dir = r"E:\Violence_Datasets\real life violence situations\Real Life Violence Dataset//"


def video_slicer(name, path, sep_frames=30):
    print('\n', "---------------------start---------------------")
    path_2 = path + "//" + name
    vid_cap = cv2.VideoCapture(path_2)
    name_2 = name.split('.')[0]  
    fps = vid_cap.get(cv2.CAP_PROP_FPS)
    size = (int(vid_cap.get(cv2.CAP_PROP_FRAME_WIDTH)), int(vid_cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
    frame_num = vid_cap.get(cv2.CAP_PROP_FRAME_COUNT)
    vid_num = int(frame_num // sep_frames)
    print('Name:{}, Location:{}'.format(name, path))
    print('frame_rate:{}  size:{} frames:{} cuts:{}'.format(fps, size, frame_num, vid_num))
    for i in range(vid_num):
        fourcc = cv2.VideoWriter_fourcc('m', 'p', '4', 'v')
        num = sep_frames * i
        print('\r', "slicing number{} cut".format(i+1), end='', flush=True)

        isExist = os.path.exists(path + '//' + 'slicer')
        if not isExist:
            os.makedirs(path + '//' + 'slicer')

        vid_path = path + '//' + 'slicer' + '//' + name_2 + '_{}.mp4'.format(i)
        vout = cv2.VideoWriter(vid_path, fourcc, fps, size)
        for f in range(sep_frames):
            success, frame = vid_cap.read()
            if f == sep_frames - 1:
                vout.release()
            else:
                vout.write(frame)
                # cv2.imshow('frame', frame)
                # cv2.waitKey(1)
    print('\n', "----------------------end----------------------")
    vid_cap.release()


for root, dirs, files in os.walk(vid_dir):
    for d in dirs:
        # d is the folder's name,label fight as ‘1’, label normal as '0'.
        path = vid_dir + '//' + d
        for root_2, dirs_2, files_2 in os.walk(path):
            for f in files_2:
                video_slicer(f, path, sep_frames=30)

```

### CSV maker

```
import csv
import os

vid_dir = r"E://ALL_dataset//"  # The name of the path to be added, the subfolder is the tag name
name = "ALL_fight_vid.csv"  # Name of the dataset

a = os.walk(vid_dir)
csv_file = open(name, "w")
writer = csv.writer(csv_file)
writer.writerow(['video', 'label'])
found = False

for root, dirs, files in os.walk(vid_dir):
    for d in dirs:
        # d is the name of the folder, the fight case is indicated by label 1, and the normal case is indicated by label 0.
        if d == 'fight':
            label = 1
        elif d == 'normal':
            label = 0
        path = vid_dir + d
        for root_2, dirs_2, files_2 in os.walk(path):
            for f in files_2:
                path_2 = path + "//" +f
                csv_read = open(name, "r")  # Reset the search pointer to the beginning of the csv file
                for line in csv_read:
                    if path_2 in line:
                        found = True
                        print('found')
                if not found:
                    writer.writerow([path_2, label])
                    print('not found')
                found = False
```

**Comment by SKBL5694** [comment]
@jwhmicrsoft I will try your code later. Thanks a lot, you are so kind. :)

**Comment by Jonyond-lin** [comment]
> btw, I found that cv2.resize is much better than the skimage's resize which makes the training much much slower. I highly recommend replace it by cv2.resize. here is my code：
> 
> ### Loading the video data
> ```
> class VideoDataset(Dataset):
> 
>     def __init__(self, datas, timestep=28, rgb=3, h=120, w=180):
>         self.datalocation = datas
>         self.timesep, self.rgb, self.h, self.w = timestep, rgb, h, w
> 
>     def __len__(self):
>         return len(self.datalocation)
> 
>     def __getitem__(self, idx):
>         if torch.is_tensor(idx):
>             idx = idx.tolist()
>         video = capture(self.datalocation.iloc[idx, 0], self.timesep, self.rgb, self.h, self.w)
>         # sample = {'video': torch.from_numpy(video), 'label': self.datalocation.iloc[idx, 1]}
> 
>         return torch.from_numpy(video), self.datalocation.iloc[idx, 1]
> ```
> 
> ### Capture and resize the frame of the video
> ```
> 
> def capture(filename, timesep, rgb, h, w):
>     tmp = []
>     frames = np.zeros((timesep, rgb, h, w), dtype=np.float32)
>     i = 0
>     vc = cv2.VideoCapture(filename)
>     frame = np.zeros((1, h, w), dtype=np.float32)
>     frame_all = vc.get(cv2.CAP_PROP_FRAME_COUNT) 
>     # frames = np.zeros((frame_all, rgb, h, w), dtype=np.float32) 
>     if vc.isOpened():
>         rval, frame = vc.read()
> 
>     frm = cv2.resize(frame, (w, h))
>     frm = np.expand_dims(frm, axis=0)
>     frm = np.moveaxis(frm, -1, 1)
> 
>     if (np.max(frm) > 1):
>         frm = frm / 255.0
>     frames[i][:] = frm
>     i += 1
> 
>     while i < timesep and i < frame_all:
>         tmp[:] = frm[:]
>         rval, frame = vc.read()
>         frm = cv2.resize(frame, (w, h))
>         frm = np.expand_dims(frm, axis=0)
>         if np.max(frm) > 1:
>             frm = frm / 255.0
>         frm = np.moveaxis(frm, -1, 1)
>         frames[i - 1][:] = frm  # - tmp
>         i += 1
>     return frames
> ```

Hello! May I have your train code if it's convenient? I need to deal with  a binary classication task in frames which is very similar to yours, but I am confused about the label generation method. Looking forward to your reply~

**Comment by Jonyond-lin** [comment]
> Btw, I wrote a csv maker and a video slicer, hope you will find it helpful.
> 
> ### Video Slicer
> ```
> 
> import cv2
> import os
> 
> vid_dir = r"E:\Violence_Datasets\real life violence situations\Real Life Violence Dataset//"
> 
> 
> def video_slicer(name, path, sep_frames=30):
>     print('\n', "---------------------start---------------------")
>     path_2 = path + "//" + name
>     vid_cap = cv2.VideoCapture(path_2)
>     name_2 = name.split('.')[0]  
>     fps = vid_cap.get(cv2.CAP_PROP_FPS)
>     size = (int(vid_cap.get(cv2.CAP_PROP_FRAME_WIDTH)), int(vid_cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
>     frame_num = vid_cap.get(cv2.CAP_PROP_FRAME_COUNT)
>     vid_num = int(frame_num // sep_frames)
>     print('Name:{}, Location:{}'.format(name, path))
>     print('frame_rate:{}  size:{} frames:{} cuts:{}'.format(fps, size, frame_num, vid_num))
>     for i in range(vid_num):
>         fourcc = cv2.VideoWriter_fourcc('m', 'p', '4', 'v')
>         num = sep_frames * i
>         print('\r', "slicing number{} cut".format(i+1), end='', flush=True)
> 
>         isExist = os.path.exists(path + '//' + 'slicer')
>         if not isExist:
>             os.makedirs(path + '//' + 'slicer')
> 
>         vid_path = path + '//' + 'slicer' + '//' + name_2 + '_{}.mp4'.format(i)
>         vout = cv2.VideoWriter(vid_path, fourcc, fps, size)
>         for f in range(sep_frames):
>             success, frame = vid_cap.read()
>             if f == sep_frames - 1:
>                 vout.release()
>             else:
>                 vout.write(frame)
>                 # cv2.imshow('frame', frame)
>                 # cv2.waitKey(1)
>     print('\n', "----------------------end----------------------")
>     vid_cap.release()
> 
> 
> for root, dirs, files in os.walk(vid_dir):
>     for d in dirs:
>         # d is the folder's name,label fight as ‘1’, label normal as '0'.
>         path = vid_dir + '//' + d
>         for root_2, dirs_2, files_2 in os.walk(path):
>             for f in files_2:
>                 video_slicer(f, path, sep_frames=30)
> ```
> 
> ### CSV maker
> ```
> import csv
> import os
> 
> vid_dir = r"E://ALL_dataset//"  # The name of the path to be added, the subfolder is the tag name
> name = "ALL_fight_vid.csv"  # Name of the dataset
> 
> a = os.walk(vid_dir)
> csv_file = open(name, "w")
> writer = csv.writer(csv_file)
> writer.writerow(['video', 'label'])
> found = False
> 
> for root, dirs, files in os.walk(vid_dir):
>     for d in dirs:
>         # d is the name of the folder, the fight case is indicated by label 1, and the normal case is indicated by label 0.
>         if d == 'fight':
>             label = 1
>         elif d == 'normal':
>             label = 0
>         path = vid_dir + d
>         for root_2, dirs_2, files_2 in os.walk(path):
>             for f in files_2:
>                 path_2 = path + "//" +f
>                 csv_read = open(name, "r")  # Reset the search pointer to the beginning of the csv file
>                 for line in csv_read:
>                     if path_2 in line:
>                         found = True
>                         print('found')
>                 if not found:
>                     writer.writerow([path_2, label])
>                     print('not found')
>                 found = False
> ```
Btw, may I ask a small question about how can we deal with the videos in our dataset whose length vary a lot?

**Comment by mamonraab** [comment]
guys i have shifted my focus about 1yr  on new network family  that i created and i named it DEVTR , it have achvied very high results , also i make as library you can check it from here  https://github.com/mamonraab/Data-efficient-video-transformer
https://ieeexplore.ieee.org/abstract/document/9530829
https://pypi.org/project/menovideo/

also i have created new version of it and now will be presented in icivc conference that will add so many features to the library once update to V2
